### PR TITLE
Expose Data Processor widget to python

### LINKED ISSUE #14732: Build DataProcessorAlgorithm Framework
A DataProcessorAlgorithm batch mode framework is required. As discussed in the [design document](https://github.com/mantidproject/documents/blob/master/Design/DataProcessorAlgorithmUI/DataProcessingUserInterface.md#implemenation-steps) (now approved). The best way to deliver this functionality would be to refactor what we already have.

This issue is concerned with building the new framework step-wise by refactoring the existing UI.

### PR DESCRIPTION
**Pull request by raquelalvarezbanos**
Exposes the `DataProcessor` widget to Python and adds an example interface illustrating how to insert the widget in a Python interface. When reviewers are happy with this PR I will remove the python interface from the menu (but will keep the code) and resolve merge conflicts.

**To test:**

- It would be good if someone familiar with the `DataProcessor` UI reviewed code changes. I had to rewrite some code as there were methods (mainly those involving unique pointers) which I didn't know how to export. In addition, some methods were rewritten to return/take `QString`s instead of `std::string`s and `QSet` instead of `std::set` (otherwise those should be exported to Python too).
- It would be good if someone familiar with the reflectometry GUI tested that functionality has not changed:
  - All actions should work as before, both from the menus and from the toolbar.
  - Error/warning messages should be shown as before (for instance warnings when processing only one row in a group, etc).
  - Runs can be transferred from the search panel (for `OFFSPEC` try with the usual `1520249`, for `INTER` try with `1620340`).
  - The interface should still be able to reduce different types of runs (i.e. single-period and multi-period).
  - Any other functionality related to the `Runs` tab you can think of.
- Test the python interface (`Interfaces` > `Utility` > `DataProcessorInterface`).

Fixes #14732.

Release notes: internal change, no need to mention this in the release notes.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by raquelalvarezbanos** [comment]
@PranavBahuguna you're probably the best person to test the reflectometry GUI (the polref one). Please make sure that nothing has changed in terms of functionality (actions, processing of runs, population of columns after reduction, rounding options, error messages, etc). If you have time to have a look at the new python interface introduced in this PR (can be accessed from `Interfaces` > `Utility` > `DataProcessorInterface`), that would be great. Test as much as you can (actions, error/warning messages, reduction, etc).

**Comment by raquelalvarezbanos** [comment]
FYI: @OwenArnold @AntonPiccardoSelg @soininen

File: scripts/Interface/ui/dataprocessorinterface/data_processor_gui.py [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        """
+        Executes an action
+        """
+        command.execute()
+
+    def _run_python_code(self, text):
+        """
+        Re-emits 'runPytonScript' signal
+        """
+        mantidplot.runPythonScript(text, True)
```
Comment (OwenArnold) [inline scripts/Interface/ui/dataprocessorinterface/data_processor_gui.py]: Looks good. We've already discussed the possibility of adding another  toy example.

File: scripts/Interface/ui/CMakeLists.txt [hunk 2]
```diff
@@ -24,6 +25,7 @@ add_custom_target(CompileUIUI DEPENDS
   CompileUIInElastic
   CompileUIReflectometer
   CompileUISANS
+  CompileUIDataProcessorInterface
```
Comment (OwenArnold) [inline scripts/Interface/ui/CMakeLists.txt]: I would suggest not calling this dataprocessorinterface. What you have written is not meant to be used in production yet and is specific to reflectometry.

I'm also not sure about shipping this. Maybe we should keep it out of the main distirbution? @martyngigg thoughts?

Comment (raquelalvarezbanos) [inline scripts/Interface/ui/CMakeLists.txt]: Another option is to ship the code but make the interface invisible to users. Just let me know what is best. Regarding the name, I am open to suggestions.

**Comment by PranavBahuguna** [comment]
@raquelalvarezbanos Apologies for the late review, but I have finished testing the `Runs` tab in the `ISIS Reflectometry` GUI. I have found several issues both related and unrelated to this PR:

* Related issues:
  - Not necessarily a bug. After logging into ICAT and searching runs using `1620340` with `INTER`, I noticed that after transferring them the order the groups were displayed in was noticeably different from that of `master`.
  - The `Instrument` selection on the bottom panel of the interface tab is initially blank, with the only option being all instrument names connected together. Attempting to process with that name selected will fail and throw an error. I also noticed that using the top `Instrument` selection, switching from `INTER` to another instrument and back will also set the connected name.
  - Using invalid properties in the options column (e.g. WavelengthMin=4, WavelengthMax=2) does not show an error message in the results log (although it fails correctly).

* Unrelated issues (occurs on master too):
  - When opening a new table, clicking 'no' or 'close' on the `Open Table?` dialog will bring up the dialog  a 2nd time.
  - Using `Import .TBL` but closing the dialog without importing anything still brings up the `Open Table?` dialog.
  - Selecting the `Output Notebook `option then attempting to process a run throws an unexpected exception.

**Comment by raquelalvarezbanos** [comment]
Thanks @PranavBahuguna. Regarding the issues you've found:

> * Related issues:
>   - Not necessarily a bug. After logging into ICAT and searching runs using `1620340` with `INTER`, I noticed that after transferring them the order the groups were displayed in was noticeably different from that of `master`.

This is expected due to a change in PR https://github.com/mantidproject/mantid/pull/19082

>  - The `Instrument` selection on the bottom panel of the interface tab is initially blank, with the only option being all instrument names connected together. Attempting to process with that name selected will fail and throw an error. I also noticed that using the top `Instrument` selection, switching from `INTER` to another instrument and back will also set the connected name.

I can't reproduce this, for me the combo box looks correct. What interface are you testing? The reflectometry GUI or the toy GUI I added in this PR?

>  - Using invalid properties in the options column (e.g. WavelengthMin=4, WavelengthMax=2) does not show an error message in the results log (although it fails correctly).

I don't think the interface will throw an error, but as you say, the algorithm will. This is the behaviour I expect.

> * Unrelated issues (occurs on master too):
>   - When opening a new table, clicking 'no' or 'close' on the `Open Table?` dialog will bring up the dialog  a 2nd time.

Ok, this is something we should fix, could you open an issue and fix it at some point before the next release? It should be very minor.

>  - Using `Import .TBL` but closing the dialog without importing anything still brings up the `Open Table?` dialog.
>  - Selecting the `Output Notebook `option then attempting to process a run throws an unexpected exception.

I don't seem to be able to reproduce these two, let's have a look together.

File: scripts/Interface/ui/CMakeLists.txt (hunk 2 above)
Comment (martyngigg) [inline scripts/Interface/ui/CMakeLists.txt]: I suppose I would consider what the main purpose of it is. Is it to provide an example of how to do this for users? If so then should we put it in to some kind of examples directory?

Comment (raquelalvarezbanos) [inline scripts/Interface/ui/CMakeLists.txt]: I'd say the main purpose is to provide an example for developers.

**Comment by AntonPiccardoSelg** [comment]
@PranavBahuguna , @OwenArnold  I guess I will take over this PR, since I will be needing the functionality soon. Just to get me up to speed: besides fixing the conflicts and getting the tests passing, where there any further requested changes?

**Comment by OwenArnold** [comment]
@AntonPiccardoSelg - No additional requests from me.

**Comment by AntonPiccardoSelg** [comment]
@PranavBahuguna I looked at the merge conflicts and they are mostly related to changes to the `getProcessingOptions`, `getPreprocessingOptions` and the old `getPreprocessingOptionsAsString` methods which is something that was changed in one of your latest PRs. The changes are quite significant and I am afraid that I might break your work when trying to resolve these issues. Could you have a look at this please (or alternatively resolve this with me)?

**Comment by martyngigg** [comment]
I assume we can move this back to `3.11`?

**Pull request merged**